File: src/dataset1/heap_sort.py
import os
import time
from helper import read_file
from csv import DictWriter
from visualize import plot_timing_data

def print_array(arr):
    print(" ".join(map(str, arr)))

def write_array_to_file(arr, filename):
    os.makedirs(os.path.dirname(filename), exist_ok=True)
    with open(filename, 'w') as file:
        file.write(" ".join(map(str, arr)) + "\n")

def heapify(arr, n, i):
    largest = i
    left = 2 * i + 1
    right = 2 * i + 2
    
    if left < n and arr[left] > arr[largest]:
        largest = left
    
    if right < n and arr[right] > arr[largest]:
        largest = right
    
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)

def heappush(heap, item):
    heap.append(item)
    current = len(heap) - 1
    while current > 0:
        parent = (current - 1) >> 1
        if heap[current] > heap[parent]:
            heap[current], heap[parent] = heap[parent], heap[current]
            current = parent
        else:
            break

def heappop(heap):
    n = len(heap)
    if n == 0:
        return None
    if n == 1:
        return heap.pop()
    
    top = heap[0]
    heap[0] = heap.pop()
    heapify(heap, len(heap), 0)
    return top

def heap_sort(arr):
    heap = []
    # Measure time to insert all data into the heap
    start_insert = time.time()
    for element in arr:
        heappush(heap, element)
    end_insert = time.time()
    insert_time = (end_insert - start_insert) * 1_000  # Convert to millisecond

    # Measure time to dequeue all data from the heap
    start_dequeue = time.time()
    sorted_arr = []
    while heap:
        sorted_arr.append(heappop(heap))
    end_dequeue = time.time()
    dequeue_time = (end_dequeue - start_dequeue) * 1_000  # Convert to millisecond

    return sorted_arr, insert_time, dequeue_time

def main():
    timings = {}
    sizes = [100, 1000, 10000, 100000, 500000, 1000000]
    insert_times = []
    dequeue_times = []
    for i, size in enumerate(sizes, start=1):
        dataset = read_file(f"sets/{i}.txt")
        print(f"Set {i} (Before sorting):")
        print_array(dataset)
        
        dataset, insert_time, dequeue_time = heap_sort(dataset)
        timings.update({f"{size}": insert_time + dequeue_time})



        print(f"\nTime taken to insert data into the heap for Set {i}: {insert_time:.0f} millisecond")
        print(f"Time taken to dequeue (remove) data from the heap for Set {i}: {dequeue_time:.0f} millisecond\n")
        
        print(f"Set {i} (After sorting):")
        print_array(dataset)

        # Write the sorted array to a file
        output_filename = f"sorted_sets/sorted_set_{i}.txt"
        write_array_to_file(dataset, output_filename)
        print(f"Sorted dataset {i} stored in '{output_filename}'\n")

        insert_times.append(insert_time)
        dequeue_times.append(dequeue_time)

    with open("sorted_sets/timings.csv", 'w') as timing:
        writer = DictWriter(timing, fieldnames=list(timings.keys()))
        writer.writeheader()
        writer.writerow(timings)


if __name__ == '__main__':
    main()
    # Call the function with your CSV file
    plot_timing_data('sorted_sets/timings.csv')

File: src/heap_sort.js
const fs = require('fs');
const { performance } = require('perf_hooks');
const { createObjectCsvWriter } = require('csv-writer');
const { writeFileSync, mkdirSync } = require('fs');
const { join } = require('path');

function printArray(arr) {
    // Prints the elements of the array in a single line
    console.log(arr.join(" "));
}

function writeArrayToFile(arr, filename) {
    // Creates the directory if it does not exist and writes the array to a file
    // mkdirSync(path.dirname(filename), { recursive: true });
    writeFileSync(filename, arr.join(" ") + "\n");
}

function heapify(arr, n, i) {
    // Maintains the heap property between parent and children nodes
    let largest = i;
    let left = 2 * i + 1;
    let right = 2 * i + 2;

    if (left < n && arr[left] > arr[largest]) {
        largest = left;
    }

    if (right < n && arr[right] > arr[largest]) {
        largest = right;
    }

    if (largest !== i) {
        [arr[i], arr[largest]] = [arr[largest], arr[i]];
        heapify(arr, n, largest);
    }
}

function heappush(heap, item) {
    // Adds an item to the heap while maintaining the heap property
    heap.push(item);
    let current = heap.length - 1;
    while (current > 0) {
        let parent = (current - 1) >> 1;
        if (heap[current] > heap[parent]) {
            [heap[current], heap[parent]] = [heap[parent], heap[current]];
            current = parent;
        } else {
            break;
        }
    }
}

function heappop(heap) {
    // Removes and returns the top element of the heap, re-balancing it afterwards
    let n = heap.length;
    if (n === 0) {
        return null;
    }
    if (n === 1) {
        return heap.pop();
    }

    let top = heap[0];
    heap[0] = heap.pop();
    heapify(heap, heap.length, 0);
    return top;
}

function heapSort(arr) {
    // Sorts an array using the heap sort algorithm and measures the time taken for insertion and removal
    let heap = [];
    // Measure time to insert all data into the heap
    let startInsert = performance.now();
    for (let element of arr) {
        heappush(heap, element);
    }
    let endInsert = performance.now();
    let insertTime = (endInsert - startInsert);

    // Measure time to dequeue all data from the heap
    let startDequeue = performance.now();
    let sortedArr = [];
    while (heap.length > 0) {
        sortedArr.push(heappop(heap));
    }
    let endDequeue = performance.now();
    let dequeueTime = (endDequeue - startDequeue);

    return { sortedArr, insertTime, dequeueTime };
}

function readFileSync(fileName) {
    return fs.readFileSync(fileName, 'utf8').trim().split(/\s+/).map(Number);
}

async function main() {
    let timings = {};
    let sizes = [100, 1000, 10000, 100000, 500000, 1000000];
    let insertTimes = [];
    let dequeueTimes = [];

    for (let i = 0; i < sizes.length; i++) {
        let size = sizes[i];
        let dataset = readFileSync(join(__dirname, `sets/${i + 1}.txt`));
        console.log(`Set ${i + 1} (Before sorting):`);
        printArray(dataset);

        let { sortedArr, insertTime, dequeueTime } = heapSort(dataset);
        timings[size] = insertTime + dequeueTime;

        console.log(`\nTime taken to insert data into the heap for Set ${i + 1}: ${insertTime.toFixed(0)} milliseconds`);
        console.log(`Time taken to dequeue (remove) data from the heap for Set ${i + 1}: ${dequeueTime.toFixed(0)} milliseconds\n`);

        console.log(`Set ${i + 1} (After sorting):`);
        printArray(sortedArr);

    

        // Write the sorted array to a file
        let outputFilename = join(__dirname, `sorted_sets/sorted_set_${i + 1}.txt`);
        writeArrayToFile(sortedArr, outputFilename);
        console.log(`Sorted dataset ${i + 1} stored in '${outputFilename}'\n`);

        insertTimes.push(insertTime);
        dequeueTimes.push(dequeueTime);
    }

    const csvWriter = createObjectCsvWriter({
        path: join(__dirname, 'sorted_sets/timings.csv'),
        header: Object.keys(timings).map(key => ({ id: key, title: key }))
    });

    await csvWriter.writeRecords([timings]);
    console.log('Timing data saved to sorted_sets/timings.csv');
}

main();
